### PR TITLE
Feature: Generative art toolbar 

### LINKED ISSUE #9737: Generative art toolbar
Since there is no way of fullscreen, seeing it on a new tab, and also resetting the animation without reloading the page, there is a drop toolbar 

## design
- updated in small things handoff: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?type=design&node-id=1287-18959&mode=design&t=5apKDpdFtsISv8TI-11
- icons change to k-grey on hover 
- container stays the same, only thing that animates are the icons 
- on desktop it appears when you hover over the img 
- when the reload button is clicked, loader is displayed instead of the item until its possible to click again (use the same we are already using somewhere)

image
image

ignore the gallery item and mobile mockups - it is an older version of the design - the important thing is the button placement 

image

## tooltips 
- it will show small tooltips when hovering over the icon

like this: 
image

image

## Mobile
on touchdevices, there is no hover of the img, so for that case, the toolbar is put underneath the img 

image

- toolbar is then stretched to the full width with icons centered

### PR DESCRIPTION
**Pull request by Temidayo32**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

- [x] Closes #9737
- [x] Related #9355 

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [ ] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality

#### Optional

- [ ] I've tested it at </ksm/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Did your issue had any of the "$" label on it?

- [ ] Fill up your DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer/?target=14rs1iTduAn9YrqtDwsTZxK9YDbkzDekBstqxc78M6An87fX&usdamount=0)

#### Community participation

- [x] [Are you at KodaDot Ecosystem Telegram?](https://t.me/koda_eco)

## Screenshot 📸

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

image

### CODE REVIEW

**Comment by Jarsen136** [comment]
The problems mentioned in the previous PR still exist #9791

**Comment by Temidayo32** [comment]
> The problems mentioned in the previous PR still exist #9791

Is there something I am missing? I believe I have fixed all the issue.

In this [comment](https://github.com/kodadot/nft-gallery/pull/9791#issuecomment-2005096407), @prury reported I removed the **fullscreen functionality**, but I don't think that is the case as seen in the gif below.



In the same comment, it was reported the toolbar should only appear on generative items, and I think I fixed that.



In another comment, it was reported I haven't added tooltip, and I think I have fixed that issue.

**Comment by prury** [comment]
@Temidayo32 when you hover over a non-genart Nft image, a full screen button should appear on the top right, and it is not present in your solution

**Comment by Temidayo32** [comment]
> @Temidayo32 when you hover over a non-genart Nft image, a full screen button should appear on the top right, and it is not present in your solution

Oh, I didn't realise that is what you meant. I actually removed that feature. I will re-add it back.

**Comment by exezbcz** [comment]
+1, lets use it on all, sry for the confusion

**Comment by Temidayo32** [comment]
> +1, lets use it on all, sry for the confusion

Can you please clarify what you mean?

**Comment by Temidayo32** [comment]
> @Temidayo32 when you hover over a non-genart Nft image, a full screen button should appear on the top right, and it is not present in your solution

Would this still be necessary then?

**Comment by exezbcz** [comment]
@Temidayo32 nope

**Comment by Temidayo32** [comment]
> @Temidayo32 nope

Done

File: locales/en.json [hunk 1]
```diff
@@ -35,6 +35,8 @@
   "chart": "Chart",
   "holders": "Holders",
   "flippers": "Flippers",
+  "fullscreen": "Fullscreen",
+  "new tab": "Open in New Tab",
```
Comment (Jarsen136) [inline locales/en.json:39]: Could you please use Camel case? `newTab`
You do have to add translation for all the other language.

**Review by Jarsen136 — CHANGES_REQUESTED** [review]
- It does not properly handle the toolbar for other kinds of nfts
`MediaType.VIDEO, MediaType.IMAGE, MediaType.OBJECT`

- The `reload` button only works for `iframe` nft

- The `New tab` button only works for `iframe` nft

**Comment by yangwao** [comment]
okay I don't count on this one for release, waiting @prury to test it through

**Comment by prury** [comment]
> okay I don't count on this one for release, waiting @prury to test it through

I will only have access to a monitor in 30m

File: components/gallery/GalleryItem.vue [hunk 2]
```diff
@@ -241,25 +254,13 @@ const tabs = {
...
+
 const hasResources = computed(
   () => nftResources.value && nftResources.value?.length > 1,
 )
-const hasAnimatedResources = computed(
-  () =>
-    nftResources.value &&
-    nftResources.value?.length > 1 &&
-    nftResources.value[1].animation,
-)
```
[the remark this answers is not in the document]
Comment (Temidayo32) [reply components/gallery/GalleryItem.vue:262]: The computed value is only relevant for the preview feature that is now removed (given the comments above).

File: components/gallery/GalleryItem.vue [hunk 3]
```diff
@@ -241,25 +254,13 @@ const tabs = {
 }
 const activeTab = ref(tabs.activity)
 
-const canPreview = computed(
-  () =>
-    !nftMimeType.value ||
-    [MediaType.VIDEO, MediaType.IMAGE, MediaType.OBJECT].includes(
-      resolveMedia(nftMimeType.value),
-    ),
-)
```
[the remark this answers is not in the document]
Comment (Temidayo32) [reply components/gallery/GalleryItem.vue:250]: Same applies here.

File: locales/en.json (hunk 1 above)
Comment (Temidayo32) [reply locales/en.json]: Done

**Comment by Temidayo32** [comment]
> [MediaType.VIDEO, MediaType.IMAGE, MediaType.OBJECT].includes(
>       resolveMedia(nftMimeType.value),
>     ),
> )

I am looking into this

**Comment by prury** [comment]
So, after discussing, the bar should appear on every gallery item across app.
after checking i still see items without it, be it gen art or not

examples: 

https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/96-9
https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/110-22
https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/91-22

**Comment by prury** [comment]
checked videos and they work fine, reloading button resets the video properly

not sure about what to do regarding audio files: 

https://deploy-preview-9809--koda-canary.netlify.app/ksm/gallery/15297998-900d19dc7d3c444e4c-PLANETPO-DEEPOLOGIC_PLANET_PO_EP-00000004

wdyt @exezbcz, disable fullscreen for them?

**Comment by exezbcz** [comment]
> checked videos and they work fine, reloading button resets the video properly
> 
> not sure about what to do regarding audio files:
> 
> [deploy-preview-9809--koda-canary.netlify.app/ksm/gallery/15297998-900d19dc7d3c444e4c-PLANETPO-DEEPOLOGIC_PLANET_PO_EP-00000004](https://deploy-preview-9809--koda-canary.netlify.app/ksm/gallery/15297998-900d19dc7d3c444e4c-PLANETPO-DEEPOLOGIC_PLANET_PO_EP-00000004)
> 
> wdyt @exezbcz, disable fullscreen for them?

full screen is oki, reload does not make much sense

File: components/gallery/GalleryItem.vue [hunk 4]
```diff
@@ -345,6 +353,68 @@ function toggleFullscreen() {
...
 }
 
+function handleReloadClick() {
+  isLoading.value = true
+
+  const reloadElement = (selector: string, isImage?: boolean) => {
+    setTimeout(() => {
+      isLoading.value = false
+      const element: ReloadElement = document.querySelector(selector)
+      if (element) {
```
Comment (preschian) [inline components/gallery/GalleryItem.vue:363]: can we put it as one liner here?

```suggestion
      if (element && isImage) {
```

**Review by preschian — APPROVED** [review]
otherwise code lgtm

followup issue: move the toolbar to a new component

**Comment by Temidayo32** [comment]
> otherwise code lgtm
> 
> followup issue: move the toolbar to a new component

I will look out for this and work on it

File: components/gallery/GalleryItem.vue [hunk 5]
```diff
@@ -345,6 +353,68 @@ function toggleFullscreen() {
...
+  const imageType = resolveMedia(nftMimeType.value)
+
+  if ([MediaType.IFRAME].includes(mediaType)) {
+    reloadElement('iframe[title="html-embed"]')
+  } else if ([MediaType.VIDEO].includes(mediaType)) {
+    reloadElement('video[data-testid="type-video"]')
+  } else if ([MediaType.OBJECT].includes(mediaType)) {
+    reloadElement('object')
+  } else if ([MediaType.IMAGE].includes(imageType)) {
+    reloadElement('img[data-testid="type-image"]', true)
```
Comment (Jarsen136) [inline components/gallery/GalleryItem.vue:389]: Please do not rely on `testid`, it is only used for e2e testing.

File: components/gallery/GalleryItem.vue [hunk 6]
```diff
@@ -345,6 +353,68 @@ function toggleFullscreen() {
...
+      isLoading.value = false
+      const element: ReloadElement = document.querySelector(selector)
+      if (element) {
+        if (isImage) {
+          let src = element.src
+          const timestamp = new Date().getTime()
+          const separatorIndex = src.indexOf('?')
+          if (separatorIndex !== -1) {
+            src = src.substring(0, separatorIndex)
+          }
```
Comment (Jarsen136) [inline components/gallery/GalleryItem.vue:370]: I would use `searchParams`
https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams

**Review by Jarsen136 — COMMENTED** [review]
It does not load 
image

https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/106-35

File: components/gallery/GalleryItem.vue (hunk 4 above)
Comment (Temidayo32) [reply components/gallery/GalleryItem.vue:363]: I think this causes a typescript error in the `else` statement (`element is null`)

**Comment by yangwao** [comment]
hey this PR is taking too long to make and don't want to risk something will broke

image

**Comment by Temidayo32** [comment]
> It does not load image
> 
> https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/106-35

It works on my end here.

**Comment by yangwao** [comment]
> It works on my end here.

image

**Comment by yangwao** [comment]
semi works but noticed delayed loading?

https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/91-161

**Comment by Temidayo32** [comment]
> semi works but noticed delayed loading?
> 
> https://deploy-preview-9809--koda-canary.netlify.app/ahp/gallery/91-161

I'm not sure

**Comment by yangwao** [comment]
Thanks, closing because it eats too much time on value it adds, thank you a lot one more time.
pay 50 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 9.55 USD/DOT ~ 5.236 $DOT
🧗 [14rs1iTduAn9YrqtDwsTZxK9YDbkzDekBstqxc78M6An87fX](https://polkadot.subscan.io/account/14rs1iTduAn9YrqtDwsTZxK9YDbkzDekBstqxc78M6An87fX)
🔗 [0xc0366ebe7fed52921658ce883987a020793013f6d12d9eba8ca618068b990b9d](https://polkadot.subscan.io/extrinsic/0xc0366ebe7fed52921658ce883987a020793013f6d12d9eba8ca618068b990b9d)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**